cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/store/SourceControlMetadataStore.java

```diff
@@ -0,0 +1,146 @@
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package io.cdap.cdap.internal.app.store;
+
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.collect.ImmutableList;
+import io.cdap.cdap.common.ApplicationNotFoundException;
+import io.cdap.cdap.proto.id.ApplicationId;
+import io.cdap.cdap.proto.id.ApplicationReference;
+import io.cdap.cdap.proto.sourcecontrol.SourceControlMeta;
+import io.cdap.cdap.spi.data.StructuredRow;
+import io.cdap.cdap.spi.data.StructuredTable;
+import io.cdap.cdap.spi.data.StructuredTableContext;
+import io.cdap.cdap.spi.data.TableNotFoundException;
+import io.cdap.cdap.spi.data.table.field.Field;
+import io.cdap.cdap.spi.data.table.field.Fields;
+import io.cdap.cdap.spi.data.table.field.Range;
+import io.cdap.cdap.store.StoreDefinition;
+import java.io.IOException;
+import java.time.Instant;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.List;
+import java.util.Optional;
+import javax.annotation.Nullable;
+
+public class SourceControlMetadataStore {
+  private StructuredTable sourceControlMetadataTable;
+  private final StructuredTableContext context;
+
+  public static SourceControlMetadataStore create(StructuredTableContext context) {
+    return new SourceControlMetadataStore(context);
+  }
+
+  private SourceControlMetadataStore(StructuredTableContext context) {
+    this.context = context;
+  }
+
+  private StructuredTable getSourceControlMetadataTable() {
+    try {
+      if (sourceControlMetadataTable == null) {
+        sourceControlMetadataTable = context.getTable(
+            StoreDefinition.SourceControlMetadataStore.SOURCE_CONTROL_METADATA);
+      }
+    } catch (TableNotFoundException e) {
+      throw new RuntimeException(e);
+    }
+    return sourceControlMetadataTable;
+  }
+
+  public void updateAppScmMeta(ApplicationId appId, SourceControlMeta scmMeta)
+      throws IOException, ApplicationNotFoundException {
+    StructuredTable scmTable = getSourceControlMetadataTable();
+    List<Field<?>> fields = getPrimaryKey(appId);
+    Optional<StructuredRow> existing = scmTable.read(fields);
+    if (!existing.isPresent()) {
+      throw new ApplicationNotFoundException(appId);
+    }
+    fields.add(Fields.stringField(StoreDefinition.SourceControlMetadataStore.SPECIFICATION_HASH_FIELD,
+        scmMeta == null ? "" : scmMeta.getFileHash()));
+    fields.add(Fields.stringField(StoreDefinition.SourceControlMetadataStore.COMMIT_ID_FIELD,
+        scmMeta == null ? "" : scmMeta.getCommitId()));
+    fields.add(Fields.longField(StoreDefinition.SourceControlMetadataStore.LAST_MODIFIED_FIELD,
+        scmMeta == null ? 0L : scmMeta.getLastSyncedAt().toEpochMilli()));
+    scmTable.upsert(fields);
+  }
+
+  public void setAppSourceControlMeta(ApplicationId appId, SourceControlMeta sourceControlMeta)
```

Comment:
can `sourceControlMeta` be `null`? If yes, please add a `nullable` annotation.

Please add javadocs for all public methods. This comment applies to whole PR.